src/antsibull/announcements.py

```diff
@@ -0,0 +1,305 @@
+        except ClientResponseError as exc:
+            eprint(f"Failed to retrieve data for ansible=={ansible_version}: {exc}")
+            return None
+        try:
+            dists = release.get_sdist_and_wheel()
+        except ValueError as exc:
+            eprint(exc)
+            return None
+        if dist_dir and (err := await verify_dists(dists, dist_dir)):
+            eprint(err)
+            return None
+
+        version = dependency_data.ansible_version
+        major_version = PypiVer(dependency_data.ansible_version).major
+        core_version = dependency_data.ansible_core_version
+        core_version_obj = PypiVer(dependency_data.ansible_core_version)
+        core_major_version = f"{core_version_obj.major}.{core_version_obj.minor}"
+        build_data_path = f"{BUILD_DATA_URL}/blob/{version}/{major_version}"
+        ctx = TemplateVars(
+            version=version,
+            major_version=major_version,
+            core_version=core_version,
+            core_major_version=core_major_version,
+            build_data_path=build_data_path,
+            release_tarball=dists.sdist,
+            release_wheel=dists.wheel,
+        )
+        return ctx
+
+
+async def _announcements_command(
+    ansible_version: str,
+    output_dir: Path,
+    dist_dir: Path | None,
+    dependency_data: DependencyFileData,
+    send: bool,
+) -> int:
+    if not (ctx := await get_data(ansible_version, dist_dir, dependency_data)):
+        return 1
+    for path in write_announcements(ANNOUNCEMENTS, ctx, output_dir):
+        print("Wrote:", path)
+    data = AnnouncementsInfo(template_vars=ctx, outputs=list(ANNOUNCEMENTS))
+    info_path = output_dir / "announcements.json"
+    write_announcements_json(data, info_path)
+    print("Wrote:", info_path)
+    if send and (rc := _send_announcements_command(output_dir, set(ACTIONS), True)):
+        return rc
+    return 0
+
+
+def write_announcements_json(info: AnnouncementsInfo, file: StrOrBytesPath) -> None:
+    with open(file, "w") as fp:
+        # pyre-ignore[16] # pyre doesn't understand pydantic's __getattr__ magic
+        json.dump(info, fp, default=pydantic.json.pydantic_encoder, indent=2)
+
+
+def load_announcements_json(file: StrOrBytesPath) -> AnnouncementsInfo:
+    with open(file) as fp:
+        data = json.load(fp)
+    return AnnouncementsInfo(**data)
+
+
+@lru_cache
+def get_body(directory: Path, name: str) -> str:
+    return (directory / name).read_text()
+
+
+def forum_announcement_webbrowser(
+    directory: Path,
+    info: AnnouncementsInfo,
+    ctx: SendCtx,  # pylint: disable=unused-argument
+) -> None:
+    subject = SUBJECT.format(info.template_vars["version"])
+    body = get_body(directory, "ansible-email-announcement.txt")
+    params_dict = FORUM_PARAMS | dict(title=subject, body=body)
+    params = "?" + urlencode(params_dict, quote_via=url_quote)
+    url = urljoin(ANSIBLE_FORUM_URL, "new-topic") + params
```

Comment:
I think I saw aiohttp somewhere in the deps. This means that you have yarl in the env always present. Instead of playing with strings and urlencoding, use that. Maybe even convert the constant from string so that everything would get a pathlib-like interface globally.

Reply:
Sure, I can try that.

Reply:
@webknjaz, I'm trying to replace this with

```
    params_dict: dict[str, str] = {"subject": subject, "body": body}
    url = URL("mailto:" + ",".join(MAIL_RECIPIENTS)) % params_dict
    webbrowser.open(str(url))

```

but it's not working as expected. It's escaping spaces as `+` while I want it to encode them as `%20` like `quote_via=urllib.parse.quote` does. Mail clients -- well, at least Thunderbird -- just leaves verbatim `+` signs which is definitely not what I want.

Reply:
Do you have an example URI you're going for? It seems weird to me me that there's many emails. I'd expect them in CC.

Reply:
The messages are sent to the ansible-project, ansible-devel, and ansible-announce lists, but that's not the problem here. The issues is the way in which the URL parameters are being encoded.

Reply:
Could you still show me what it does vs. what you expect?

Reply:
I want

```
mailto:ansible-devel@googlegroups.com,ansible-project@googlegroups.com,ansible-announce@googlegroups.com?subject=Release%20announcement%3A%20Ansible%20community%20package%209.2.0&body=Hello%20everyone
```

as opposed to

```
mailto:ansible-devel@googlegroups.com,ansible-project@googlegroups.com,ansible-announce@googlegroups.com?subject=Release+announcement:+Ansible+community+package+9.2.0&body=Hello+everyone
```